reduce/src/prep.py

```diff
@@ -0,0 +1,356 @@
+        "build_command",
+        help="The specific build command preamble used sans the target. "
+        "Example: `make -j$(nproc) LLVM=1 V=1`",
+        nargs="*",
+        default=["make", f"-j{os.cpu_count()}", "LLVM=1", "V=1"],
+    )
+
+    parser.add_argument(
+        "target",
+        help="The source file you wish to target. Example: lib/string.o",
+        type=validate_target_from_cli,
+    )
+
+    parser.add_argument(
+        "-p",
+        "--path-to-linux",
+        type=lambda arg: validate_path_from_cli(arg, "-p/--path-to-linux"),
+        default="./",
+        help="Specify the Linux source tree directory",
+    )
+
+    parser.add_argument(
+        "-f",
+        "--force-rm-existing-target",
+        help="If the target (foo.o) already exists, remove it -- without prompting",
+        action="store_true",
+        default=False,
+    )
+
+    parser.add_argument(
+        "-F",
+        "--force-rm-existing-script",
+        help="If test.sh already exists, remove it -- without prompting",
+        action="store_true",
+        default=False,
+    )
+
+    parser.add_argument(
+        "-d",
+        "--debug",
+        help="Enable prepreduce Debug logs",
+        action="store_true",
+        default=False,
+    )
+
+    parser.add_argument(
+        "-o",
+        "--output",
+        help="Output directory of prepreduce outputs (.i files and test scripts)",
+        default=os.getcwd(),
+        type=lambda arg: validate_path_from_cli(arg, "-o/--output", allow_mkdir=True),
+    )
+
+
+def validate_cli_args(cli_args):
+    DEBUG = cli_args.debug
+    if DEBUG:
+        ic.enable()
+
+    validate_build_cmd_from_cli(cli_args.build_command)
+
+
+def is_kernel_configured_for_clang(path_to_linux: Path) -> Optional[bool]:
+    """
+    Find out what compiler to use based on what the kernel build system is
+    currently configured for
+    """
+
+    config_file = Path(path_to_linux) / Path(".config")
+    if not config_file.exists():
+        error(
+            f"No .config file found at {config_file}, please configure your "
+            "kernel build or \n"
+            "change your linux build path with --path-to-linux /path/to/linux"
+        )  # intentional newline in above error
+
+    clang_kernel_config_opt = "CONFIG_CC_IS_CLANG=y"
+    gcc_kernel_config_opt = "CONFIG_CC_IS_GCC=y"
+    with config_file.open() as fd:
+        for line in fd:
+            # option is typically found *very* early in the file,
+            # yielding an inexpensive linear search
+            line = line.strip()
+            if clang_kernel_config_opt in line:
+                return True
+            if gcc_kernel_config_opt in line:
+                return False
+
+    raise EOFError(
+        "Upon parsing .config, could not determine gcc or clang usage. "
+        f"reduce prep is specifically looking for `{clang_kernel_config_opt}` "
+        f"or `{gcc_kernel_config_opt}` in your {config_file} file"
+    )
```

Comment:
I think this could be simplified to just:

```python
text = config_file.read_text(encoding='utf-8')
return "CONFIG_CC_IS_CLANG=y" in text
```

The kernel only supports GCC and clang, so if `CONFIG_CC_IS_CLANG` is not `y`, we have to be using GCC:

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/scripts/Kconfig.include#n44

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/scripts/cc-version.sh

Reply:
Opted for this approach to specifically avoid reading the entire `.config` file which could be pretty large.

Reply:
I noticed one of those two config options always in the first 15 lines of `.config` so I'd figure I 'd check for either or.